docs: add more context to vim.schedule_wrap

rktjmp opened the pull request:
The docs for `schedule` and `schedule_wrap` do not clearly explain why you would use one vs the other. By expanding the `schedule_wrap` docs, its intended usage is hopefully clearer.

- Explicitly state that `schedule_wrap` returns a function in the one line summary.
- Remove the usage of the term "defer" to avoid confusion with `vim.defer_fn` which calls `vim.schedule_wrap` internally.
- Mention that arguments are passed along.
- Include some example usage.
- Rename param from `cb` to `callback` to match vim.schedule

rktjmp committed:
docs: clarify vim.schedule_wrap behaviour

- Remove the usage of the term "defer" to avoid confusion with
  `vim.defer_fn`, which also calls `vim.schedule_wrap` internally.
- Explicitly state that `vim.schedule_wrap` returns a function in the
  text.
- Mention that arguments are passed along.
- Include a usage example.
- Rename param to `fn`.

bfredl commented on runtime/doc/lua.txt:
```diff
@@ -1714,11 +1714,43 @@ vim.region({bufnr}, {pos1}, {pos2}, {regtype}, {inclusive})
+        local function my_function(event)
+          -- Some "unsafe" code ...
+        end
+        vim.api.nvim_create_autocmd("TextChangedI", {
```
autocmd is not the best example here as the API is already safe in an autocmd. Deferring autocmds can be useful, but _not_ for the reason stated in the comment just above, which makes this confusing. An example with `vim.uv` callbacks would be better right here.

rktjmp replied:
How about

```lua
function notify_readable(_err, readable)
  -- Without schedule_wrap, this would raise the error "nvim_echo must not be called in a lua loop callback"
  vim.notify("readable? " .. tostring(readable))
end
vim.uv.fs_access(vim.fn.stdpath("config"), "R", vim.schedule_wrap(notify_readable))
```

- fs_access is pretty understandable without knowing much about uv
- Passes an argument to callback
- Not much pre-work fanfare (eg, no other handles need to be created etc)

A timer is used in other examples, but that does not pass a value back to the callback which I think is helpful to include.

justinmk commented on runtime/doc/lua.txt:
```diff
@@ -1714,11 +1714,43 @@ vim.region({bufnr}, {pos1}, {pos2}, {regtype}, {inclusive})
+    Returns a function which calls {callback} via |vim.schedule()|.
+
+    Useful to "wrap" existing functions, ensuring they are only called when
+    the Nvim API is safe to call. The returned function passes all arguments
```
This is redundant with the `|vim.schedule()|` docs, which are already linked. Any part of this that isn't covered there should be added to the `|vim.schedule()|` docs.

rktjmp replied:
> Returns a function which calls {callback} via |vim.schedule()|.
> 
> The returned function passes all arguments to {callback}.

?

lewis6991 replied:
```vimdoc
vim.schedule({callback})                                      *vim.schedule()*
    Schedules {callback} to be invoked soon by the main event-loop. Useful to
    avoid |textlock| or other temporary restrictions.

    Parameters:  
      • {callback}  fun()
```

rktjmp replied:
Not sure I understand your intent Lewis.

lewis6991 replied:
Sorry, I misunderstood your comment. Ignore mine.

justinmk commented on runtime/doc/lua.txt:
```diff
@@ -1714,11 +1714,43 @@ vim.region({bufnr}, {pos1}, {pos2}, {regtype}, {inclusive})
+        -- The following functions are equivalent.
+        -- In both cases, the functions effectively become non-blocking.
+        local function manually_scheduled(arg)
+          vim.schedule(function()
```
Example is way too verbose.

justinmk commented:
Off-topic for this PR, but we should not have 3 functions for these features. The features of `defer_fn` and `schedule_wrap` could be part of `schedule()`, assuming they are needed at all. defer_fn appears to be yet another way to start a timer, while saving the caller the trouble of calling schedule_wrap?

lewis6991 commented:
> Off-topic for this PR, but we should not have 3 functions for these features. The features of `defer_fn` and `schedule_wrap` could be part of `schedule()`, assuming they are needed at all. defer_fn appears to be yet another way to start a timer, while saving the caller the trouble of calling schedule_wrap?

Bear in mind for any consolidation we do in this area we should avoid introducing any polymorphic returns types as they often make the code harder to type and generally reason about.

justinmk commented on runtime/lua/vim/_editor.lua:
```diff
@@ -316,18 +316,29 @@ do
+---@param callback function
 ---@return function
-function vim.schedule_wrap(cb)
+function vim.schedule_wrap(callback)
```
Going to insist on "fn" since I don't want to see "callback" proliferate. Functions are just functions, usually.

```suggestion
function vim.schedule_wrap(fn)
```

rktjmp replied:
100%. Can I also change `vim.schedule`?

justinmk commented on runtime/lua/vim/_editor.lua:
```diff
@@ -316,18 +316,29 @@ do
+--- function notify_readable(_err, readable)
+---   vim.notify("readable? " .. tostring(readable))
+--- end
+--- vim.uv.fs_access(vim.fn.stdpath("config"), "R", vim.schedule_wrap(notify_readable))
```
This is a good example.

rktjmp committed:
docs: update vim.schedule param name and type

Per https://github.com/neovim/neovim/pull/25286#discussion_r1332861721
and https://github.com/neovim/neovim/pull/25286#discussion_r1334318352

lewis6991 commented:
Can you squash the commits and update the commit message.

zeertzjq commented on runtime/lua/vim/_meta/builtin.lua:
```diff
@@ -194,10 +194,10 @@ function vim.str_utfindex(str, index) end
 --- to avoid |textlock| or other temporary restrictions.
---- @param callback fun()
-function vim.schedule(callback) end
+--- @param fn fun()
```
```suggestion
--- @param fn function
```

rktjmp replied:
Seems to be a few of those around:

```
runtime/lua/vim/treesitter/languagetree.lua|187 col 4| ---@param f fun(): R1, R2, R2
runtime/lua/vim/_system.lua|235 col 5| --- @param on_error fun()
runtime/lua/vim/_system.lua|247 col 4| ---@param cb fun()
runtime/lua/vim/_meta/builtin.lua|199 col 5| --- @param fn fun()
runtime/lua/vim/_meta/builtin.lua|235 col 5| --- @param callback? fun(): boolean Optional callback. Waits until {callback} returns true
runtime/lua/vim/_meta/builtin.lua|283 col 5| --- @param callback fun()
runtime/lua/vim/_meta/misc.lua|13 col 5| --- @param func fun()
```

Want me to patch them all?

lewis6991 replied:
No. A lot/all of those are intentional and are used by LuaLs.
